### PR TITLE
Export and merge functions

### PR DESCRIPTION
**Pull request by ajparsons**
Repackaged version of the export feature branch - split into more commits to separate different features and refactoring. Incorporating changes made to package since December. 

Now also includes a feature for intelligent merging of similar popolo (where there will be overlap of individuals - but one file might have additional memberships). 

Includes workaround for no to_string function in approx_date package - this logic also in https://github.com/mysociety/approx_dates/pull/2 to approx_dates.

Tests written for export + merge features.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,6 @@
 /htmlcov
 /everypolitician_popolo.egg-info
 /dist
+.project
```
Comment (mhl) [inline .gitignore:10]: I think generally I'd expect people to put ignore rules that are specific to a particular IDE into `.git/info/exclude` (or to have them in a per-user gitignore configured with: `git config --global core.excludesfile $HOME/.gitignore`. (I'm happy with adding the `.gitattributes` line-endings configuration, though.)

File: popolo_data/collections.py [hunk 2]
```diff
@@ -8,9 +8,11 @@
 
 class PopoloCollection(object):
 
-    def __init__(self, data_list, object_class, all_popolo):
+    object_class = None
+
+    def __init__(self, data_list , all_popolo):
         self.all_popolo = all_popolo
-        self.object_class = object_class
+        self.object_class = self.__class__.object_class
```
Comment (mhl) [inline popolo_data/collections.py:16]: I don't think this line is actually necessary, is it? `self.foo` falls back to trying to find a class variable foo, e.g.:

```
In [1]: class Thing(object):
   ...:     foo = 'class foo'
   ...:     bar = 'class bar'
   ...:     def __init__(self):
   ...:         self.foo = 'instance foo'
   ...:     def print_stuff(self):
   ...:         print self.foo
   ...:         print self.bar
   ...:         

In [2]: Thing().print_stuff()
instance foo
class bar
```

File: popolo_data/base.py [hunk 3]
```diff
@@ -4,44 +4,20 @@
 
 
 from approx_dates.models import ApproxDate
-from six.moves.urllib_parse import urlsplit
```
Comment (mhl) [inline popolo_data/base.py:7]: With a refactoring like this commit, I'd really like the tests to still pass at this point, but the imports in other modules haven't been updated and the tests don't pass at this point, so I think the commit is incomplete. (That would also make some of the later commits easier to read.)

File: popolo_data/importer.py [hunk 4]
```diff
@@ -19,35 +23,22 @@ def from_url(cls, url):
...
-
-    @property
-    def events(self):
-        return EventCollection(self.json_data.get('events', []), self)
+    @classmethod
+    def new(cls):
+        return cls({})
 
+    def __init__(self, json_data=None):
+        if json_data == None:
```
Comment (mhl) [inline popolo_data/importer.py]: Tiny stylistic thing - since `None` is a singleton, PEP 8 suggests `if json_data is None:`

File: popolo_data/base.py [hunk 5]
```diff
@@ -1,30 +1,180 @@
-from datetime import date
```
Comment (mhl) [inline popolo_data/base.py:1]: There are some typos in this commit message (e.g. "In new system", "a attribute", extra space before "to self.data"

File: popolo_data/base.py [hunk 6]
```diff
@@ -1,30 +1,180 @@
...
+    id = Attribute(default="16")
+    
+    Sets a default extraction value of 16.
+    
+    id = Attribute(null=True)
+    
+    will return None rather than an error if the value is absent.
+    
+    '''
+    def __init__(self,attr="",default=None,null=False,allow_multiple=False):
```
Comment (mhl) [inline popolo_data/base.py:70]: There should be spaces after the commas here (we really should have an automated check for this kind of thing)

File: popolo_data/base.py [hunk 7]
```diff
@@ -1,30 +1,180 @@
...
+        self.attr = attr
+        self.default_value = default
+        self.allow_null_default = null
+        self.allow_multiple = allow_multiple
+
+    def __get__(self, obj, type=None):
+        if self.allow_null_default == False and self.default_value == None:
```
Comment (mhl) [inline popolo_data/base.py:84]: Couldn't this be `if (not self.allow_null_default) and (self.default_value is None):` ?

**Review by mhl — CHANGES_REQUESTED** [review]
Hi @ajparsons - I think I'm OK with these changes in principle, but there are few things that make this hard to review for me. I've added HoundCI now, who's automatically commented on the not very interesting stylistic things. I've also left some line comments (although GitHub showing some of these as "outdated" now after I've rebased).

A key thing is that I'd like the tests to pass at each commit; I mentioned this particularly in the case of a commit that should have just been a refactoring, but caused the tests to start failing. A nice trick for checking that this is the case is to use:

```
git rebase -i --exec 'tox' master
```

So basically, I like the direction this is heading in, but I'd like to come back to it once some of these things are sorted out.
